### PR TITLE
Patch MXNet to use a thread-local engine

### PR DESCRIPTION
**Pull request by hqucms**
For testing in https://github.com/cms-sw/cmssw/pull/23768.

The original `NaiveEngine` in MXNet is not thread safe. Here we change the engine to be a thread local object, so it will allow each thread to use a different engine:
https://github.com/hqucms/incubator-mxnet/commit/f31352f26b816d8a2c7080f6fc81377ff4924369

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

for a test by itself

**Comment by slava77** [comment]
instead of pulling from https://hqu.web.cern.ch/hqu/tools/mxnet, shouldn't we use a CMS/SW repo?

**Comment by smuzaffar** [comment]
+externals

File: mxnet-predict.spec [hunk 1]
```diff
@@ -1,7 +1,7 @@
-### RPM external mxnet-predict 1.2.0
+### RPM external mxnet-predict 1.2.1.mod1
 %define projectname mxnet-predict
 %define releasename apache-mxnet-src-%{realversion}-incubating
-Source: https://github.com/apache/incubator-mxnet/releases/download/%{realversion}/apache-mxnet-src-%{realversion}-incubating.tar.gz
+Source: https://hqu.web.cern.ch/hqu/tools/mxnet/apache-mxnet-src-%{realversion}-incubating.tar.gz
```
Comment (davidlange6) [inline mxnet-predict.spec]: hi- we should use a GitHub fork in the cms-sw area instead of a random (likely short lived) web address for this source

Comment (slava77) [reply mxnet-predict.spec]: makes sense.
@smuzaffar  do we have a fork for mxnet external under cms-sw ?

Comment (davidlange6) [reply mxnet-predict.spec]: We do not - lets make one - is the patch above in a repo someplace?

**Comment by davidlange6** [comment]
hold

File: mxnet-predict.spec (hunk 1 above)
Comment (hqucms) [reply mxnet-predict.spec]: The tarball I made is based on https://github.com/apache/incubator-mxnet/releases/tag/1.2.1 and the patch at https://github.com/hqucms/incubator-mxnet/commit/f31352f26b816d8a2c7080f6fc81377ff4924369

Comment (mrodozov) [reply mxnet-predict.spec]: I can't fork repos on cms-externals. If you fork it for me, I have the rest prepared.

**Comment by slava77** [comment]
@mrodozov @davidlange6 
please clarify the status of making a fork repo for this update on cms-externals

**Comment by mrodozov** [comment]
@slava77 I don't have permissions to make new repos on cms-externals, as stated in the comments. If you fork it for me I'll take it from there

**Comment by slava77** [comment]
I'm guessing that  @smuzaffar is still away.
we could probably wait or see if @fabiocos or @davidlange6 can make that repo/fork in cms-externals

**Comment by slava77** [comment]
@fabiocos @davidlange6 @smuzaffar 
please clarify when we can get the repo/fork in cms-externals

**Comment by davidlange6** [comment]
Hi @mrodozov
does this work ?

https://github.com/cms-externals/incubator-mxnet

**Comment by hqucms** [comment]
@kpedro88 found another thread-safety problem with MXNet, and it seems that we need an extra patch on the MXNet source code. https://github.com/cms-sw/cmsdist/pull/4207/commits/517a8a4b3f3b2dc29bf12c38cd48f3ae8687db3a switches to this new patched version. The changes w.r.t to the official MXNet v1.2.1 can be found at: https://github.com/apache/incubator-mxnet/compare/1.2.1...hqucms:1.2.1-mod.

**Comment by mrodozov** [comment]
@davidlange6 it will work yes.

**Comment by mrodozov** [comment]
unhold

**Comment by davidlange6** [comment]
unhold

**Comment by davidlange6** [comment]
hi @slava77 - this PR does not require cms-sw/cmssw/#23768, right? (eg, only the reverse is true)

**Comment by slava77** [comment]
On 8/14/18 3:15 PM, David Lange wrote:
> hi @slava77 - this PR does not require cms-sw/cmssw/#23768, right? (eg,
> only the reverse is true)

right

**Comment by davidlange6** [comment]
merge

**Pull request merged**